### PR TITLE
Break out "meta" property into separate object

### PR DESCRIPTION
**Pull request by sselberg**
Fixes eiffel-community/eiffel#256

### Applicable Issues
https://github.com/eiffel-community/eiffel/issues/256

### Description of the Change
Breaking out "meta" as a separate versioned object.
Changing:
* Documentations
* Schemas
* Adds ref support to the schema validation script.

### Alternate Designs
I haven't identified any alternate designs.

### Possible Drawbacks
Not identified.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Sven Selberg sven.selberg@axis.com

### CODE REVIEW

**Comment by raja-maragani** [comment]
@sselberg Changes are good. 

These Eiffel schemas are used in https://github.com/eiffel-community/eiffel-remrem-semantics application.
The REMReM semantics application is generating the Java pojo classes using by the Json schema's. 
Now here we modified the Json schema's, then it will be good to test the **REMReM semantics Pojo generation and unit tests**.

jsonschema2pojo: https://github.com/joelittlejohn/jsonschema2pojo

**Comment by sselberg** [comment]
@raja-maragani unfortunately I haven't been able to generate pojos even with eiffel:edition-agen-1 that is specifed in eiffel-remrem-sematics:master (0390ff2) /.pom.xml. It blows up when running test for EiffelActivityFinishedEvent:

```
eiffel-remrem-semantics (master)$ proxy mvn clean install -P process-schemas
...
...
eiffelenvironmentdefinedevent.json
eiffelsourcechangesubmittedevent.json
eiffeltestsuitestartedevent.json
eiffelactivityfinishedevent.json
14:47:59.721 [main] ERROR com.ericsson.eiffel.remrem.semantics.validator.EiffelValidator - Cannot parse JSON schema. The resource: schemas/input/EiffelActivityFinishedEvent.json. class java.lang.IllegalArgumentException:  argument "in" is null
java.lang.IllegalArgumentException: argument "in" is null
	at com.fasterxml.jackson.databind.ObjectMapper._assertNotNull(ObjectMapper.java:4757)
	at com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:3022)
	at com.ericsson.eiffel.remrem.semantics.validator.EiffelValidator.<init>(EiffelValidator.java:81)
	at com.ericsson.eiffel.remrem.semantics.validator.OutputValidator.<init>(OutputValidator.java:24)
	at com.ericsson.eiffel.remrem.semantics.factory.EiffelOutputValidatorFactory.<clinit>(EiffelOutputValidatorFactory.java:36)
	at com.ericsson.eiffel.remrem.semantics.SemanticsService.outputValidate(SemanticsService.java:280)
	at com.ericsson.eiffel.remrem.semantics.SemanticsService.generateMsg(SemanticsService.java:238)
	at com.ericsson.eiffel.remrem.semantics.SemanticsService.generateMsg(SemanticsService.java:193)
	at com.ericsson.eiffel.remrem.semantics.ServiceTest.testGenerate(ServiceTest.java:97)
```

File: schemas/EiffelMetaProperty/1.0.0.json [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+          "EiffelIssueVerifiedEvent",
+          "EiffelSourceChangeCreatedEvent",
+          "EiffelSourceChangeSubmittedEvent",
+          "EiffelTestCaseCanceledEvent",
+          "EiffelTestCaseFinishedEvent",
+          "EiffelTestCaseStartedEvent",
+          "EiffelTestCaseTriggeredEvent",
+          "EiffelTestExecutionRecipeCollectionCreatedEvent",
+          "EiffelTestSuiteFinishedEvent",
+          "EiffelTestSuiteStartedEvent"
```
Comment (magnusbaeck) [inline schemas/EiffelMetaProperty/1.0.0.json:35]: To make the protocol extensible we shouldn't enumerate that valid event types here. A simple regexp pattern would do.

File: examples/validate.py [hunk 2]
```diff
@@ -91,14 +91,22 @@ def validateExamples(examples, schemas, maxExamples, shuffle):
...
   latestReportTime = time.perf_counter()
-  
+  # Keys in schemas are $TYPE-$VERSION.
+  # All files are at ./$TYPE/$VERSION.json.
+  # Links from file to other file are always ../$OTHER_TYPE/$VERSION.json.
+  schema_store = {
+    "../" + key.replace("-", "/") + ".json": schemas[key] for key in schemas
+  }
+  resolver = RefResolver(None, referrer=None, store=schema_store)
+
```
Comment (magnusbaeck) [inline examples/validate.py:101]: This feels a bit icky. If we create a RefResolver with a base URI (i.e. first positional argument) that matches the path to the original schema it'll load the EiffelMetaProperty by itself. I replaced line 108, below, with
```python
        validator = Draft4Validator(
          schemas[schemaKey],
          resolver=RefResolver(
            "file://" + os.path.join(
              os.getcwd(),
              "schemas/EiffelCompositionDefinedEvent/foo",
            ),
            None,
          ),
        )
```
and it worked fine. The "foo" part is obviously bogus and should be the actual path to the schema file.

We should probably replace the whole `schemas` dict with a dict that maps from (type, version) tuples to validator instances. I think that would improve the code in general.

**Review by magnusbaeck — COMMENTED** [review]
I think this is a great start! I made a few comments and I'll try this out with a couple of other tools that perform validation or otherwise read the schema files. They'll probably need some changes to resolve the referenced schemas, and that's fine, but since none of us (I think) are JSON schema experts we should make sure that examples/validate.py doesn't just _happen_ to work.

Obviously we'd have to issue new versions of all event types rather than change the most recent version.

If we can move forward with this I'd like to apply the same treatment to (primarily) the links and probably a few other common substructures too (like logs and gitIdentifier).

File: examples/validate.py (hunk 2 above)
Comment (sselberg) [reply examples/validate.py:101]: It would make total sense to keep the validators in a collection instead of the current schema dict.
The validators would have to share resolver (with all available schemas) since we don't know what other schema files, the schema we are using to validate, references. Unless we want to enumerate the "sub-schema" used (only meta in this case but in the future potentially more) and populate that list when we add a new "sub-schema".

**Comment by m-linner-ericsson** [comment]
We might also want to try our schema if they work with https://github.com/quicktype/quicktype, see https://app.quicktype.io/ to test it

File: schemas/EiffelMetaProperty/1.0.0.json [hunk 3]
```diff
@@ -0,0 +1,147 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "meta": {
+    "type": "object",
+    "properties": {
```
[the remark this answers is not in the document]
Comment (sselberg) [reply schemas/EiffelMetaProperty/1.0.0.json]: Done

File: schemas/EiffelMetaProperty/1.0.0.json (hunk 1 above)
Comment (sselberg) [reply schemas/EiffelMetaProperty/1.0.0.json]: Done

File: examples/validate.py (hunk 2 above)
Comment (sselberg) [reply examples/validate.py]: Done

**Comment by m-linner-ericsson** [comment]
As this PR is so big I would avoid using force push. It becomes very hard to know what happened between two commits. 

For the record: @magnusbaeck I know you use force push. What is your take on using force push on a PR of this size?

**Comment by sselberg** [comment]
> As this PR is so big I would avoid using force push. It becomes very hard to know what happened between two commits.
> 

It's been a while since I worked in GitHub but didn't GitHub used to save a reference to previous states of the PR?
Or is that only when the merge-branch is in the same repository?

File: examples/validate.py [hunk 4]
```diff
@@ -66,15 +66,15 @@ def loadAllJsonObjects(dir):
...
-  schemas = {}
-  for path, fileName, o in schemaTuples:
-    schemaName = os.path.basename(os.path.dirname(path))
-    versionName = fileName[:-5]
-    schemas[schemaName + "-" + versionName] = o
+def loadValidators():
+  schemas, badSchemaFiles = loadAllJsonObjects("schemas")
+  validators = {}
+  for path, fileName, o in schemas:
+    key = os.path.basename(os.path.dirname(path)) + "-" + fileName[:-5]
```
Comment (m-linner-ericsson) [inline examples/validate.py:73]: Could we get back some names, like this?
```suggestion
   schemaName = os.path.basename(os.path.dirname(path))
   versionName = fileName[:-5]
   key = schemaName + "-" + versionName
```

The names make it easier to read what is happening.

**Review by m-linner-ericsson — CHANGES_REQUESTED** [review]
Please note that some comments would apply to several files even though the comment is only given in one.

File: eiffel-vocabulary/EiffelMetaProperty.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
+<!---
```
Comment (m-linner-ericsson) [inline eiffel-vocabulary/EiffelMetaProperty.md:1]: I think we should update https://github.com/eiffel-community/eiffel/blob/master/eiffel-syntax-and-usage/the-meta-object.md and instead link to that one from each event.

File: eiffel-syntax-and-usage/the-meta-object.md [hunk 6]
```diff
@@ -18,7 +18,7 @@
 # The Meta Object
```
Comment (m-linner-ericsson) [inline eiffel-syntax-and-usage/the-meta-object.md:18]: As stated in the other file we should update this one instead.

File: eiffel-syntax-and-usage/the-meta-object.md [hunk 7]
```diff
@@ -18,7 +18,7 @@
...
 The __meta__ object contains meta-information describing the event: when it was created, where it came from, its type et cetera. The __meta__ object contains the same members regardless of __meta.type__<sup>[1](#footnote1)</sup>, with the caveat that certain members are optional and the tendency to use them may vary with event type.
 
-Even though the meta object is kept universal for all event types, it may change over time as the Eiffel protocol evolves. Due to the [per-event versioning](./versioning.md) of the protocol, this means that different unique combinations of event type and version may feature different meta object syntax. Consequently, the syntax is described in the documentation of each respective event type.
+Even though the meta object is kept universal for all event types, it may change over time as the Eiffel protocol evolves. Due to the [per-event versioning](./versioning.md) of the protocol, this means that different unique combinations of event type, version and meta object version may feature different meta object syntax.
 
 &nbsp;
 &nbsp;
```
Comment (m-linner-ericsson) [inline eiffel-syntax-and-usage/the-meta-object.md:24]: The footer need to be updated also

File: schemas/EiffelMetaProperty/1.0.0.json [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      "pattern": "^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$"
+    },
+    "type": {
+      "type": "string",
+      "pattern": "^Eiffel[a-zA-Z]*Event$"
+    },
+    "version": {
+      "type": "string",
+      "pattern": "^[1-9][0-9]*\\.[0-9]+\\.[0-9]+",
+      "default": "3.0.0"
```
Comment (m-linner-ericsson) [inline schemas/EiffelMetaProperty/1.0.0.json:16]: Should we really have a default value?

File: schemas/EiffelMetaProperty/1.0.0.json [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
+  "properties": {
+    "id": {
+      "type": "string",
+      "pattern": "^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$"
+    },
+    "type": {
+      "type": "string",
+      "pattern": "^Eiffel[a-zA-Z]*Event$"
```
Comment (m-linner-ericsson) [inline schemas/EiffelMetaProperty/1.0.0.json:11]: As a comment: As we no longer validate for actual event names, we should step the schema version. This is creating a weaker validation possibility. Can we send in parameters for this?

**Review by m-linner-ericsson — CHANGES_REQUESTED** [review]

File: eiffel-vocabulary/EiffelActivityCanceledEvent.md [hunk 10]
```diff
@@ -53,131 +53,8 @@ __Legal targets:__ [EiffelFlowContextDefinedEvent](../eiffel-vocabulary/EiffelFl
...
-__Format:__  
-__Required:__ Yes  
-__Description:__ The name of the sequence. There MUST not be two identical __meta.security.sequenceProtection.sequenceName__ values in the same event.
-
-##### meta.security.sequenceProtection.position
-__Type:__ Integer  
-__Format:__  
-__Required:__ Yes  
-__Description:__ The number of the event within the named sequence.
+## Meta
```
[the remark this answers is not in the document]
Comment (sselberg) [reply eiffel-vocabulary/EiffelActivityCanceledEvent.md:56]: We are referencing the"meta" attribute on an eiffel event not the attributes of the meta object.

File: eiffel-vocabulary/EiffelActivityCanceledEvent.md [hunk 11]
```diff
@@ -53,131 +53,8 @@ __Legal targets:__ [EiffelFlowContextDefinedEvent](../eiffel-vocabulary/EiffelFl
...
+See [Meta property description](./EiffelMetaProperty.md)
```
[the remark this answers is not in the document]
Comment (sselberg) [reply eiffel-vocabulary/EiffelActivityCanceledEvent.md:57]: Wouldn't that be more suitable in the actual meta object description?

File: eiffel-syntax-and-usage/the-meta-object.md (hunk 7 above)
Comment (sselberg) [reply eiffel-syntax-and-usage/the-meta-object.md:24]: What needs to be updated in the footer?

File: eiffel-syntax-and-usage/the-meta-object.md (hunk 6 above)
Comment (sselberg) [reply eiffel-syntax-and-usage/the-meta-object.md:18]: Done

File: schemas/EiffelMetaProperty/1.0.0.json (hunk 8 above)
Comment (sselberg) [reply schemas/EiffelMetaProperty/1.0.0.json:16]: This is copy/paste from the schema of the individual events. Do you want to change it in this PR?

File: schemas/EiffelMetaProperty/1.0.0.json (hunk 9 above)
Comment (sselberg) [reply schemas/EiffelMetaProperty/1.0.0.json:11]: I haven't found a way to express this with json-schema.

File: eiffel-syntax-and-usage/the-meta-object.md (hunk 7 above)
Comment (m-linner-ericsson) [reply eiffel-syntax-and-usage/the-meta-object.md:151]: I thought it didn't match what we have now. On second thought do you feel we need a footer or should we inline it with the introduction text?

Comment (sselberg) [reply eiffel-syntax-and-usage/the-meta-object.md:151]: If you're asking me I don't have a strong opinion about what the footer text should say or whether it should be moved to the main clause or not.

File: schemas/EiffelMetaProperty/1.0.0.json (hunk 9 above)
Comment (m-linner-ericsson) [reply schemas/EiffelMetaProperty/1.0.0.json:11]: @magnusbaeck I think you wanted to have a regexp here instead of listing all events. Isn't it safer to list the exact names as we now don't validate the names at all `EiffelActyvilyStratlyEvent` would be valid?

File: schemas/EiffelMetaProperty/1.0.0.json (hunk 1 above)
Comment (m-linner-ericsson) [reply schemas/EiffelMetaProperty/1.0.0.json]: I am not that sure when thinking about it. Please see my comment below.

File: schemas/EiffelMetaProperty/1.0.0.json (hunk 8 above)
Comment (m-linner-ericsson) [reply schemas/EiffelMetaProperty/1.0.0.json:16]: It would make sense when describing each event on its own but it feels strange that the default for all events should be 3.0

File: eiffel-syntax-and-usage/the-meta-object.md (hunk 7 above)
Comment (m-linner-ericsson) [reply eiffel-syntax-and-usage/the-meta-object.md:151]: I would vote for moving it.

File: eiffel-syntax-and-usage/the-meta-object.md [hunk 12]
```diff
@@ -18,7 +18,134 @@
 # The Meta Object
 The __meta__ object contains meta-information describing the event: when it was created, where it came from, its type et cetera. The __meta__ object contains the same members regardless of __meta.type__<sup>[1](#footnote1)</sup>, with the caveat that certain members are optional and the tendency to use them may vary with event type.
 
-Even though the meta object is kept universal for all event types, it may change over time as the Eiffel protocol evolves. Due to the [per-event versioning](./versioning.md) of the protocol, this means that different unique combinations of event type, version and meta object version may feature different meta object syntax.
+Even though the meta object is kept universal for all event types (i.e. all events in an edition share the same version of a meta-object, it may change over time as the Eiffel protocol evolves. Due to the [per-event versioning](./versioning.md) of the protocol, this means that different unique combinations of event type, version and meta object version may feature different meta object syntax.
```
Comment (m-linner-ericsson) [inline eiffel-syntax-and-usage/the-meta-object.md:21]: What do you think of removing line 19 and the footnote and replacing it with the following text?

```suggestion
The __meta__ object contains meta-information describing the event: when it was created, where it came from, its type et cetera. The __meta__ object contains the same members regardless of __meta.type__, with the caveat that certain members are optional and the tendency to use them may vary with event type.

Event types are versioned independently from one another as described in [per-event versioning](./versioning.md) but all events in an edition will have the same version of the meta object. 

To summarize:

1. Any change to meta requires all events to be updated.
1. Consumers should be prepared to receive events of varying meta contents between events. The exception to this are the __meta.type__ and __meta.version__ fields, which are always assumed to be present regardless of event and version.
1. Any schema of a specific version of an event must also include the linked meta object
```

File: eiffel-vocabulary/EiffelActivityCanceledEvent.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (m-linner-ericsson) [reply eiffel-vocabulary/EiffelActivityCanceledEvent.md]: Hmm, was aiming for some more information about the link and why you should follow it. Feels a bit bare right now.

**Review by m-linner-ericsson — CHANGES_REQUESTED** [review]
-

File: schemas/EiffelActivityCanceledEvent/3.0.0.json [hunk 13]
```diff
@@ -2,113 +2,7 @@
...
-      },
-      "required": [
-        "id",
-        "type",
-        "version",
-        "time"
-      ],
-      "additionalProperties": false
-    },
+    "meta": {"$ref": "../EiffelMetaProperty/1.0.0.json"},
```
[the remark this answers is not in the document]
Comment (e-backmark-ericsson) [reply schemas/EiffelActivityCanceledEvent/3.0.0.json:5]: I agree with @m-linner-ericsson. All event schemas should have their versions stepped for this change.

File: schemas/EiffelMetaProperty/1.0.0.json (hunk 8 above)
Comment (erkist) [reply schemas/EiffelMetaProperty/1.0.0.json:16]: I guess this will cause issues for libraries that want to generate events based on schema? They would all have version 3.0.0 forever. Unless we create one EiffelMetaProperty schema for each meta.version value, and reference that.

**Comment by sselberg** [comment]
See issue for details.
